crates/core/src/energy.rs

```diff
@@ -0,0 +1,127 @@
+use derive_more::{Add, AddAssign, From, Sub, SubAssign};
+use std::fmt;
+use std::time::Duration;
+
+use spacetimedb_lib::{Hash, Identity};
+
+use crate::messages::control_db::Database;
+
+/// [EnergyQuanta] represents an amount of energy in a canonical unit.
+/// It represents the smallest unit of energy that can be used to pay for
+/// a reducer invocation. We will likely refer to this unit as an "eV".
+///
+/// NOTE: This is represented by a signed integer, because it is possible
+/// for a user's balance to go negative. This is allowable
+/// for reasons of eventual consistency motivated by performance.
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Add, Sub, AddAssign, SubAssign)]
+pub struct EnergyQuanta(i128);
+
+impl EnergyQuanta {
+    pub const ZERO: Self = EnergyQuanta(0);
+
+    #[inline]
+    pub fn new(v: i128) -> Self {
+        Self(v)
+    }
+
+    #[inline]
+    pub fn get(&self) -> i128 {
+        self.0
+    }
+
+    pub fn from_disk_usage(bytes_stored: u64, storage_period: Duration) -> Self {
+        // TODO: this line is lossy if bytes_stored is >1.6 PiB. do we ever care about that case?
+        let energy = bytes_stored as f64 * storage_period.as_secs_f64();
```

Comment:
Can't multiply them otherwise. We can't just discard the fractional component of the duration, since with a large enough bytes_stored a couple of milliseconds more storage time is a non-negligible amount of energy. And, afaict, there's no way to multiply a i128 and a f64 natively/with high precision. Oh, I guess I could just do it manually - `bytes_stored / 1_000_000_000 * ns + bytes_stored * s`. (Though also to note - the potential inaccuracy of the float op would only be about 1eV, since float accuracy is in stages of powers of 2)